### PR TITLE
Create edgelet-kube - a module runtime for k8s.

### PR DESCRIPTION
**Pull request by darobs**
Create a module runtime for Kubernetes which supports the iotedged watchdog.  

Not all functions are implemented, just enough to get the watchdog to start edgeAgent.

### CODE REVIEW

File: edgelet/edgelet-kube/src/convert/to_k8s.rs [hunk 1]
```diff
@@ -1,11 +1,673 @@
...
+            name: key.clone(),
+            value: Some(val.clone()),
+            ..api_core::EnvVar::default()
+        })
+        .collect();
+    // Pass along "USE_PERSISTENT_VOLUMES" to EdgeAgent
+    if runtime.use_pvc() && EDGE_EDGE_AGENT_NAME == module_label_value {
+        let env_var = api_core::EnvVar {
+            name: USE_PERSISTANT_VOLUME_CLAIMS.to_string(),
+            value: Some("TRUE".to_string()),
```
Comment (avranju) [inline edgelet/edgelet-kube/src/convert/to_k8s.rs]: Should this be `"True"` instead?

File: edgelet/edgelet-kube/src/convert/to_k8s.rs [hunk 2]
```diff
@@ -1,11 +1,673 @@
...
+        .env()
+        .iter()
+        .map(|(key, val)| api_core::EnvVar {
+            name: key.clone(),
+            value: Some(val.clone()),
+            ..api_core::EnvVar::default()
+        })
+        .collect();
+    // Pass along "USE_PERSISTENT_VOLUMES" to EdgeAgent
+    if runtime.use_pvc() && EDGE_EDGE_AGENT_NAME == module_label_value {
```
Comment (avranju) [inline edgelet/edgelet-kube/src/convert/to_k8s.rs]: ☝️ Nice!

File: edgelet/edgelet-kube/src/convert/to_k8s.rs [hunk 3]
```diff
@@ -1,11 +1,673 @@
...
+                        volume_mounts.push(bind_mount);
+                    } else {
+                        warn!("Bind mount did not contain a source and target");
+                    }
+                }
+                Some("volume") => {
+                    if let (Some(source), Some(target)) = (mount.source(), mount.target()) {
+                        let volume_name = sanitize_dns_value(source)?;
+
+                        let volume = if runtime.use_pvc() {
```
Comment (avranju) [inline edgelet/edgelet-kube/src/convert/to_k8s.rs]: Maybe some comments here to describe what's going on might help (with the the `use_pvc` check)?

File: edgelet/edgelet-kube/src/runtime.rs [hunk 4]
```diff
@@ -14,140 +15,237 @@ use edgelet_core::{
...
+                                    .create_secret(namespace_copy.as_str(), &pull_secret)
+                                    .map_err(Error::from)
+                                    .map(|_| ());
+
+                                Either::B(f)
+                            }
+                        });
+                    Either::A(fut)
+                })
+                .unwrap_or_else(|err| Either::B(future::err(err)));
```
Comment (avranju) [inline edgelet/edgelet-kube/src/runtime.rs]: Learnt recently from Arnav that if you have a case where you have either a `FutureResult` or some other kind of `Future` to return then `Either` is not strictly necessary. For example, here on line 241 we can simply rturn `fut` and on line 243 we can convert the `Result` into a `Future` via `into_future` which could produce a `FutureResult` that yielded another `Future` (on success) which can then be _flattened_ via the `flatten` function. For example:
```rust
let fut = auth_to_image_pull_secret(self.namespace(), auth)
  .map_err(Error::from)
  .map(|(secret_name, pull_secret)| {
      let client_copy = self.client.clone();
      let namespace_copy = self.namespace().to_owned();
      self
          .client
          .lock()
          .expect("Unexpected lock error")
          .borrow_mut()
          .list_secrets(self.namespace(), Some(secret_name.as_str()))
          .map_err(Error::from)
          .and_then(move |secrets| {
              if let Some(current_secret) = secrets.items.into_iter().find(|secret| {
                  secret.metadata.as_ref().map_or(false, |meta| {
                      meta.name.as_ref().map_or(false, |n| *n == secret_name)
                  })
              }) {
                  if current_secret == pull_secret {
                      Either::A(Either::A(future::ok(())))
                  } else {
                      let f = client_copy
                          .lock()
                          .expect("Unexpected lock error")
                          .borrow_mut()
                          .replace_secret(
                              secret_name.as_str(),
                              namespace_copy.as_str(),
                              &pull_secret,
                          )
                          .map_err(Error::from)
                          .map(|_| ());

                      Either::A(Either::B(f))
                  }
              } else {
                  let f = client_copy
                      .lock()
                      .expect("Unexpected lock error")
                      .borrow_mut()
                      .create_secret(namespace_copy.as_str(), &pull_secret)
                      .map_err(Error::from)
                      .map(|_| ());

                  Either::B(f)
              }
          })
  })
  .into_future()
  .flatten();
```

File: edgelet/edgelet-kube/src/runtime.rs [hunk 5]
```diff
@@ -194,12 +292,63 @@ where
...
+                                .borrow_mut()
+                                .create_deployment(namespace_copy.as_str(), &new_deployment)
+                                .map_err(Error::from)
+                                .map(|_| ());
+                            Either::B(fut)
+                        }
+                    });
+                Either::A(fut)
+            })
+            .unwrap_or_else(|err| Either::B(future::err(err)));
```
Comment (avranju) [inline edgelet/edgelet-kube/src/runtime.rs]: As before, we should be able to use `into_future` and `flatten` and avoid an `Either`.

File: edgelet/edgelet-kube/src/convert/to_k8s.rs [hunk 6]
```diff
@@ -1,11 +1,661 @@
...
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct AuthEntry {
+    pub username: String,
+    pub password: String,
+    pub auth: String,
+}
+impl AuthEntry {
+    pub fn new(username: String, password: String) -> AuthEntry {
+        let auth = base64::encode(&format!("{}:{}", username, password));
```
[the remark this answers is not in the document]
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/to_k8s.rs]: This struct is used to serialize the data for the image pull secrets, see line 71 - that's from the image pull secrets documentation. 

The AuthEntry struct models the JSON so I can serialize it all the fields are used.  It's weird, I should document this.

File: edgelet/edgelet-kube/src/convert/to_k8s.rs [hunk 7]
```diff
@@ -1,11 +1,661 @@
...
+    // { "auths":
+    //   {"<registry>" :
+    //      { "username":"<user>",
+    //        "password":"<password>",
+    //        "email":"<email>" (not needed)
+    //        "auth":"<base64 of '<user>:<password>'>"
+    //       }
+    //   }
+    // }
+    let mut auths = BTreeMap::<String, AuthEntry>::new();
```
[the remark this answers is not in the document]
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/to_k8s.rs]: It does, removing.

File: edgelet/edgelet-kube/src/convert/to_k8s.rs (hunk 1 above)
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/to_k8s.rs]: Probably, changing it.

File: edgelet/edgelet-kube/src/convert/to_k8s.rs [hunk 8]
```diff
@@ -1,11 +1,673 @@
...
+                privileged: Some(*privileged),
+                ..api_core::SecurityContext::default()
+            };
+            Some(context)
+        } else {
+            None
+        }
+    } else {
+        None
+    };
```
[the remark this answers is not in the document]
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/to_k8s.rs]: Done!

File: edgelet/edgelet-kube/src/runtime.rs (hunk 4 above)
Comment (darobs) [reply edgelet/edgelet-kube/src/runtime.rs]: That looks a lot better, done!

File: edgelet/edgelet-kube/src/convert/to_k8s.rs (hunk 3 above)
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/to_k8s.rs]: Will do.

File: edgelet/edgelet-kube/src/convert/to_k8s.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (avranju) [reply edgelet/edgelet-kube/src/convert/to_k8s.rs]: Ah. Understood.

**Comment by avranju** [comment]
LGTM!

**Pull request merged**